src/spatial/flatten.jl

```diff
@@ -0,0 +1,90 @@
+    num_specs = size(u0, 1)
+    @assert size(hopping_constants) == size(u0) # hopping_constants[i,j] is species i, site j
+    hop_constants = Vector{Matrix{F}}(undef, size(hopping_constants, 2))
+    for site in 1:num_nodes
+        num_nbs = num_neighbors(spatial_system, site)
+        hop_constants[site] = reshape(repeat(hopping_constants[:,site], num_nbs), num_specs, num_nbs)
+    end
+    
+    flatten(netstoch, reactstoch, rx_rates, spatial_system, u0, tspan, hop_constants; kwargs...)
+end
+
+"""
+if reaction rates is a vector, assume reaction rates are equal across sites
+"""
+function flatten(netstoch::AbstractArray, reactstoch::AbstractArray, rx_rates::Vector, spatial_system, u0::Matrix{Int}, tspan, hopping_constants::Vector{Matrix{F}}; kwargs...) where F <: Number
+    num_nodes = num_sites(spatial_system)
+    rates = reshape(repeat(rx_rates, num_nodes), length(rx_rates), num_nodes)
+    flatten(netstoch, reactstoch, rates, spatial_system, u0, tspan, hopping_constants; kwargs...)
+end
+
+"""
+"flatten" the spatial jump problem. Return flattened DiscreteProblem and MassActionJump.
+"""
+function flatten(netstoch::Vector{R}, reactstoch::Vector{R}, rx_rates::Matrix{F}, spatial_system, u0::Matrix{Int}, tspan, hopping_constants::Vector{Matrix{F}}; scale_rates = true, kwargs...) where {R, F <: Number}
+    num_species = size(u0, 1)
+    num_nodes = num_sites(spatial_system)
+    num_rxs = length(reactstoch)
+    @assert length(hopping_constants) == size(u0, 2) == size(rx_rates, 2) == num_nodes
+    @assert length(netstoch) == length(reactstoch) == size(rx_rates, 1) == num_rxs
+    spec_CI = CartesianIndices((num_species, num_nodes))
+    spec_LI = LinearIndices((num_species, num_nodes))
+
+    sum_outdegrees = sum(num_neighbors(spatial_system, site) for site in 1:num_nodes)
+    num_jumps = num_species*(sum_outdegrees + num_rxs)
+    total_netstoch = R[]; sizehint!(total_netstoch, num_jumps)
+    total_reactstoch = R[]; sizehint!(total_reactstoch, num_jumps)
+    total_rates = F[]; sizehint!(total_rates, num_jumps)
+
+    #hops
+    # assuming hopping_constants isa Vector{Matrix} where hopping_constants[src][species, index] is the hop const of species from src to neighbor at index in neighbors(spatial_system, src)
+    for src in 1:num_nodes
+        hopping_constants_at_src = hopping_constants[src]
+        for (i,dst) in enumerate(neighbors(spatial_system, src))
+            for species in 1:num_species
+                dst_idx = spec_LI[spec_CI[species,dst]]
+                src_idx = spec_LI[spec_CI[species,src]]
+                push!(total_netstoch, [dst_idx => 1, src_idx => -1])
+                push!(total_reactstoch, [src_idx => 1])
+                push!(total_rates, hopping_constants_at_src[species, i])
+            end
+        end
+    end
+
+    #reactions
+    for site in 1:num_nodes
+        append!(total_netstoch, map(rx_stoch -> map(p -> Pair(spec_LI[spec_CI[p[1],site]], p[2]), rx_stoch), netstoch))
```

Comment:
Can this be written a bit more efficiently as a `foreach` that `push!` a single `map` call instead of appending two map calls?

Reply:
Again, I'm just trying to keep memory use down since it can grow quickly in flattening.